### PR TITLE
Switch checkout to nextjs routing

### PR DESCRIPTION
**Pull request by piotrgrundas**
I want to merge this change because it resolves [Switch to path-based routing and <Link> components](https://app.clickup.com/t/2549495/SALEOR-1565)
- switch checkout to nextjs,
- polish configs, libs,
- adjust service worker for nextjs.

### Pull Request Checklist

1. [x] All visible strings are translated with proper context.
1. [x] All data-formatting is locale-aware (dates, numbers, and so on).
1. [x] The changes are tested.
1. [x] The code is documented (docstrings, project documentation).
1. [x] Changes are mentioned in the changelog.

### Test Environment Config

API_URI=https://master.staging.saleor.rocks/graphql/

### CODE REVIEW

File: src/@next/hooks/useRedirectToCorrectCheckoutStep.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    totalPrice
+  );
+  const stepFromPath = CHECKOUT_STEPS.find(({ link }) => link === pathname)
+    ?.step;
+  const isShippingRequired = checkIfShippingRequiredForProducts(items);
+
+  const isIncorrectStep =
+    (pathname !== CHECKOUT_STEPS[4].link &&
+      (!stepFromPath || (stepFromPath && maxPossibleStep < stepFromPath))) ||
+    (pathname === CHECKOUT_STEPS[1].link && !isShippingRequired);
```
Comment (krzysztofwolski) [inline src/@next/hooks/useRedirectToCorrectCheckoutStep.ts:26]:
```suggestion
  const isIncorrectStep =
    !stepFromPath ||
     stepFromPath > maxPossibleStep ||
    (pathname === CHECKOUT_STEPS[1].link && !isShippingRequired);
```

Comment (piotrgrundas) [reply src/@next/hooks/useRedirectToCorrectCheckoutStep.ts:26]: 👌  I don't intend to touch existing logic not to adjacently break sth, but sure

**Comment by patrys** [comment]
Task linked: [SALEOR-1565 Switch to path-based routing and <Link> components](https://app.clickup.com/t/2549495/SALEOR-1565)

File: config/next/config.serviceWorker.js [hunk 2]
```diff
@@ -1,83 +1,121 @@
...
+        SERVICE_WORKER_URL: serviceWorkerUrl,
+      }),
     },
-    target: "serverless",
-    generateInDevMode: process.env.SERVICE_WORKER === "true",
-    sourcemap: true,
-    transformManifest: manifest => ["/"].concat(manifest),
-    workboxOpts: {
-      swDest,
-      exclude: [
```
Comment (krzysztofwolski) [inline config/next/config.serviceWorker.js:19]: There is not need for excluding those paths?

File: src/pages/_app.tsx [hunk 3]
```diff
@@ -24,6 +22,11 @@ import {
 } from "../constants";
 import { LocaleProvider } from "../components/Locale";
 import { App as StorefrontApp } from "../app";
+import { version } from "../../package.json";
+
+if (!ssrMode) {
+  window.version = version;
```
Comment (krzysztofwolski) [inline src/pages/_app.tsx:28]: 👍

File: src/@next/components/molecules/OrderTable/test.tsx [hunk 4]
```diff
@@ -129,8 +129,8 @@ describe("<OrderTable />", () => {
     expect(
       pushSpy
     ).toHaveBeenCalledWith(
-      `http://order-history/[token]?token=${token}`,
-      `/${token}`,
+      `/order-history/[token]?token=${token}`,
+      `/order-history/${token}`,
```
Comment (krzysztofwolski) [inline src/@next/components/molecules/OrderTable/test.tsx:133]: Can we use paths from src/paths.ts ? So there would be no more need for changing those in the future

File: config/next/config.serviceWorker.js (hunk 2 above)
Comment (piotrgrundas) [reply config/next/config.serviceWorker.js:19]: It's moved to l: 84

File: src/@next/components/atoms/DemoBanner/DemoBanner.tsx [hunk 5]
```diff
@@ -1,9 +1,7 @@
+import { mediumScreen } from "@styles/constants";
```
Comment (jwm0) [inline src/@next/components/atoms/DemoBanner/DemoBanner.tsx:1]: I believe the placement was correct before, storefront seems to be using the following convention:
```
<module imports>
// newline
<absolute imports>
// newline
<relative imports>
```
The eslint config should enforce that, but we can deal with that later.

File: src/@next/hooks/useRedirectToCorrectCheckoutStep.ts [hunk 6]
```diff
@@ -0,0 +1,36 @@
+import { useCart, useCheckout } from "@saleor/sdk";
+import { CHECKOUT_STEPS } from "@temp/core/config";
+import { checkIfShippingRequiredForProducts } from "@utils/core";
+import { useRouter } from "next/router";
+import { useEffect } from "react";
+import { useCheckoutStepState } from "./useCheckoutStepState";
```
Comment (jwm0) [inline src/@next/hooks/useRedirectToCorrectCheckoutStep.ts:6]: Please use the import placement convention I mentioned above

File: src/@next/hooks/useRedirectToCorrectCheckoutStep.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+  const isShippingRequired = checkIfShippingRequiredForProducts(items);
+
+  const isIncorrectStep =
+    !stepFromPath ||
+    stepFromPath > maxPossibleStep ||
+    (pathname === CHECKOUT_STEPS[1].link && !isShippingRequired);
+
+  const getStepLink = () =>
+    CHECKOUT_STEPS.find(stepObj => stepObj.step === recommendedStep)?.link ||
+    CHECKOUT_STEPS[0].link;
```
Comment (jwm0) [inline src/@next/hooks/useRedirectToCorrectCheckoutStep.ts:29]: Move this block of code inside `useEffect`

File: src/@next/pages/CheckoutPage/subpages/CheckoutAddressSubpage.tsx [hunk 8]
```diff
@@ -8,28 +8,20 @@ import React, {
...
-export interface ICheckoutAddressSubpageHandles {
-  submitAddress: () => void;
-}
-
-interface IProps extends RouteComponentProps<any> {
-  changeSubmitProgress: (submitInProgress: boolean) => void;
-  onSubmitSuccess: () => void;
-}
+import { CheckoutStep } from "@temp/core/config";
+import { SubpageBaseProps, SubpageCompleteHandler } from "../utils";
```
Comment (jwm0) [inline src/@next/pages/CheckoutPage/subpages/CheckoutAddressSubpage.tsx:19]: Please use the import placement convention I mentioned above

File: src/@next/pages/CheckoutPage/subpages/CheckoutPaymentSubpage.tsx [hunk 9]
```diff
@@ -6,34 +6,23 @@ import React, {
   useState,
 } from "react";
 import { useIntl } from "react-intl";
-import { RouteComponentProps } from "react-router";
 
 import { CheckoutPayment } from "@components/organisms";
 import { useCheckout } from "@saleor/sdk";
 import { commonMessages } from "@temp/intl";
 import { IFormError } from "@types";
+import { SubpageBaseProps, SubpageCompleteHandler } from "../utils";
```
Comment (jwm0) [inline src/@next/pages/CheckoutPage/subpages/CheckoutPaymentSubpage.tsx:14]: Please use the import placement convention I mentioned above

File: src/@next/pages/CheckoutPage/subpages/CheckoutReviewSubpage.tsx [hunk 10]
```diff
@@ -4,12 +4,13 @@ import React, {
...
   useState,
 } from "react";
-import { RouteComponentProps } from "react-router";
 
 import { CheckoutReview } from "@components/organisms";
 import { statuses as dummyStatuses } from "@components/organisms/DummyPaymentGateway";
 import { OrderStatus, useCheckout } from "@saleor/sdk";
 import { IFormError } from "@types";
+import { CheckoutStep } from "@temp/core/config";
+import { SubpageBaseProps, SubpageCompleteHandler } from "../utils";
```
Comment (jwm0) [inline src/@next/pages/CheckoutPage/subpages/CheckoutReviewSubpage.tsx:13]: Please use the import placement convention I mentioned above

File: src/@next/pages/CheckoutPage/utils.tsx [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+import type { IItems } from "@saleor/sdk/lib/api/Cart/types";
+import { checkoutMessages } from "@temp/intl";
+import {
+  CheckoutStep,
+  CheckoutStepDefinition,
+  CHECKOUT_STEPS,
+} from "@temp/core/config";
+import { checkIfShippingRequiredForProducts } from "@utils/core";
+import { paths } from "@paths";
+import { ParsedUrlQueryInput } from "querystring";
```
Comment (jwm0) [inline src/@next/pages/CheckoutPage/utils.tsx:15]: Please use the import placement convention I mentioned above

File: src/serviceWorker/index.ts [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+  CacheFirst,
+  NetworkFirst,
+  StaleWhileRevalidate,
+} from "workbox-strategies";
+import { Router } from "./router";
+import {
+  deleteEntriesForCache,
+  getBuildManifest,
+  getBuildManifestPages,
+} from "./utils";
```
Comment (jwm0) [inline src/serviceWorker/index.ts:23]: Please use the import placement convention I mentioned above

File: src/serviceWorker/router.ts [hunk 13]
```diff
@@ -0,0 +1,101 @@
+import { cacheNames, RouteHandler, RouteHandlerObject } from "workbox-core";
+import { RegExpRoute, Router as WorkboxRouter } from "workbox-routing";
+import { getRequestedPageFromURL } from "./utils";
```
Comment (jwm0) [inline src/serviceWorker/router.ts:3]: Please use the import placement convention I mentioned above

**Review by jwm0 — CHANGES_REQUESTED** [review]

File: src/@next/components/atoms/DemoBanner/DemoBanner.tsx (hunk 5 above)
Comment (piotrgrundas) [reply src/@next/components/atoms/DemoBanner/DemoBanner.tsx:1]: My vscode config got wild 🙈 ...

File: config/next/config.serviceWorker.js [hunk 14]
```diff
@@ -1,83 +1,121 @@
...
+                  .createHash("md5")
+                  .update(Buffer.from(fs.readFileSync(`public/${name}`)))
+                  .digest("hex"),
+              },
+            ];
+          }, []);
+
+        config.plugins.push(
+          new InjectManifest({
+            mode: "development",
```
[the remark this answers is not in the document]
Comment (piotrgrundas) [reply config/next/config.serviceWorker.js]: It's automatically taken from the webpack config and if I remember correctly, I tried to debug prod build. Good catch.

File: src/@next/pages/CheckoutPage/utils.tsx [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+    [CheckoutStep.Shipping]: (
+      <FormattedMessage {...checkoutMessages.shippingNextActionName} />
+    ),
+    [CheckoutStep.Payment]: (
+      <FormattedMessage {...checkoutMessages.paymentNextActionName} />
+    ),
+    [CheckoutStep.Review]: (
+      <FormattedMessage {...checkoutMessages.reviewNextActionName} />
+    ),
+  }[step]);
```
Comment (jwm0) [inline src/@next/pages/CheckoutPage/utils.tsx:65]: This object doesn't need to be created on every function call, I'd suggest moving it just above the fn itself:

```suggestion

const continueButtonTextMap: Partial<Record<CheckoutStep, JSX.Element>> = {
  [CheckoutStep.Address]: (
    <FormattedMessage {...checkoutMessages.addressNextActionName} />
  ),
  [CheckoutStep.Shipping]: (
    <FormattedMessage {...checkoutMessages.shippingNextActionName} />
  ),
  [CheckoutStep.Payment]: (
    <FormattedMessage {...checkoutMessages.paymentNextActionName} />
  ),
  [CheckoutStep.Review]: (
    <FormattedMessage {...checkoutMessages.reviewNextActionName} />
  ),
};
export const getContinueButtonText = (step: CheckoutStep) =>
  continueButtonTextMap[step];
```

File: next.config.js [hunk 16]
```diff
@@ -13,8 +14,9 @@ const withProdConfig = require("./config/next/config.prod");
 
 module.exports = withPlugins([
   [withOptimizedImages, { handleImages: ["jpeg", "png", "webp", "gif"] }],
+  withTM,
   withBaseConfig,
   withServiceWorkerConfig,
   [withDevConfig, {}, [PHASE_DEVELOPMENT_SERVER]],
-  [withProdConfig, {}, [PHASE_PRODUCTION_SERVER, PHASE_PRODUCTION_BUILD]],
+  [withProdConfig, {}, ["!" + PHASE_DEVELOPMENT_SERVER]],
```
Comment (jwm0) [inline next.config.js:21]: Is this change intentional?

Comment (piotrgrundas) [reply next.config.js:21]: Yes.

**Review by jwm0 — APPROVED** [review]

**Review by krzysztofwolski — APPROVED** [review]
